src/main/java/org/spongepowered/api/resource/Resource.java

```diff
@@ -0,0 +1,146 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.resource;
+
+import org.spongepowered.api.data.persistence.DataFormat;
+import org.spongepowered.api.data.persistence.DataFormats;
+import org.spongepowered.api.data.persistence.DataView;
+import org.spongepowered.api.resource.metadata.MetadataSection;
+import org.spongepowered.api.resource.metadata.NamedMetadataSections;
+import org.spongepowered.api.resource.pack.PackInformation;
+import org.spongepowered.api.resource.pack.PackInformationManager;
+
+import java.io.BufferedReader;
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.stream.Stream;
+
+/**
+ * A resource can represent any kind of loaded data. It can be a file on the
+ * filesystem, a network location, or even generated at runtime. Use
+ * {@link #inputStream()} to load the data held by a resource.
+ */
+public interface Resource extends Closeable {
+
+    /**
+     * Gets the path of this resource.
+     *
+     * @return The path
+     */
+    ResourcePath path();
+
+    /**
+     * Gets the name of the {@link PackInformation pack} which owns this resource. To
+     * get the instance, feed the returned value to
+     * {@link PackInformationManager#information(String)}.
+     *
+     * @return The parent pack.
+     * @see PackInformationManager#information(String)
+     */
+    String pack();
+
+    /**
+     * Returns the {@link InputStream} of this resource. Multiple calls to this
+     * method will not return a new object. To get a new object, get a new
+     * resource.
+     *
+     * @return The input stream
+     */
+    InputStream inputStream();
+
+    /**
+     * Checks whether this resource has metadata or not.
+     *
+     * @return True if this resource has metadata, false otherwise
+     */
+    boolean hasMetadata();
+
+    /**
+     * Gets the specified metadata section for this resource or
+     * {@link Optional#empty()} if it has no metadata.
+     *
+     * @param section The section serializer
+     * @return The metadata or empty if it doesn't exist
+     * @see NamedMetadataSections
+     */
+    <T> Optional<T> metadata(MetadataSection<T> section);
+
+    /**
+     * Creates a new {@link BufferedReader} from this resource's
+     * {@link InputStream}.
+     *
+     * @param charset The charset to use, usually utf-8
```

Comment:
`{@link StandardCharsets#UTF_8 utf-8}`, just to give a bit of a pointer.

Are all Mojang resources UTF-8? (I mean, I'd hope so...)